src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,1019 @@
+		D_ERROR("failed to create dataspace\n");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype, attr_dspace,
+			      H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		D_ERROR("failed to create attribute\n");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, entry->dpe_str);
+	if (status < 0) {
+		rc = -DER_IO;
+		D_ERROR("failed to write attribute "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+static int
+serialize_uint(hid_t file_id, uint64_t val, const char *prop_str)
+{
+	int	rc = 0;
+	hid_t	status = 0;
+	hsize_t	attr_dims[1];
+	hid_t	attr_dtype = 0;
+	hid_t	attr_dspace = 0;
+	hid_t	usr_attr = 0;
+
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_NATIVE_UINT64);
+	if (attr_dtype < 0) {
+		D_ERROR("failed to attribute datatype\n");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		D_ERROR("failed to create dataspace\n");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype,
+			      attr_dspace, H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		D_ERROR("failed to create attribute\n");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, &val);
+	if (status < 0) {
+		rc = -DER_IO;
+		D_ERROR("failed to write attr "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+static char*
+prop_to_str(uint32_t type)
+{
+	switch (type) {
+	case DAOS_PROP_CO_LABEL:
+		return "DAOS_PROP_CO_LABEL";
+	case DAOS_PROP_CO_OWNER:
+		return "DAOS_PROP_CO_OWNER";
+	case DAOS_PROP_CO_OWNER_GROUP:
+		return "DAOS_PROP_CO_OWNER_GROUP";
+	case DAOS_PROP_CO_ACL:
+		return "DAOS_PROP_CO_ACL";
+	case DAOS_PROP_CO_LAYOUT_TYPE:
+		return "DAOS_PROP_CO_LAYOUT_TYPE";
+	case DAOS_PROP_CO_LAYOUT_VER:
+		return "DAOS_PROP_CO_LAYOUT_VER";
+	case DAOS_PROP_CO_CSUM:
+		return "DAOS_PROP_CO_CSUM";
+	case DAOS_PROP_CO_CSUM_CHUNK_SIZE:
+		return "DAOS_PROP_CO_CSUM_CHUNK_SIZE";
+	case DAOS_PROP_CO_CSUM_SERVER_VERIFY:
+		return "DAOS_PROP_CO_CSUM_SERVER_VERIFY";
+	case DAOS_PROP_CO_REDUN_FAC:
+		return "DAOS_PROP_CO_REDUN_FAC";
+	case DAOS_PROP_CO_REDUN_LVL:
+		return "DAOS_PROP_CO_REDUN_LVL";
+	case DAOS_PROP_CO_SNAPSHOT_MAX:
+		return "DAOS_PROP_CO_SNAPSHOT_MAX";
+	case DAOS_PROP_CO_COMPRESS:
+		return "DAOS_PROP_CO_COMPRESS";
+	case DAOS_PROP_CO_ENCRYPT:
+		return "DAOS_PROP_CO_ENCRYPT";
+	case DAOS_PROP_CO_DEDUP:
+		return "DAOS_PROP_CO_DEDUP";
+	case DAOS_PROP_CO_DEDUP_THRESHOLD:
+		return "DAOS_PROP_CO_DEDUP_THRESHOLD";
+	case DAOS_PROP_CO_ALLOCED_OID:
+		return "DAOS_PROP_CO_ALLOCED_OID";
+	default:
```

Comment:
@mchaarawi maybe I missed it. Has it always been a property? It has seemed to always work even without serializing/deserializing this property. Should this property be set for all types of containers?

Reply:
well all properties were added at different points in time, but the ROOTs one has been there for a few months now.
deserializing would work, but the problem is in case of POSIX, dfs_mount should fail on the deserialized one.
the container type here is irrelevant though IMO as it could be set on any container type and it could have different meaning for the oid.

Reply:
@mchaarawi ok, I am confused why it does not seem to fail. I serialized then deserialized a container. Then I ran a "daos filesystem copy" of the deserialized container to a new container, which calls dfs_mount on the source container (i.e. the deserialized one) and it succeeds.

Reply:
looking at the code, it actually should assert:
https://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c#L1514
so im not sure what you are doing to not hit this assert. it basically checks if the property is there and asserts if it's not.

Reply:
@mchaarawi is that property not being set by dfs_cont_create? The ROOT_OID? Because all I can think is that it is getting set there when daos-deserialize calls dfs_cont_create

Reply:
ah yes it is set there if you are calling dfs cont_create.. so that's why it doesn't fail.
the problem i guess then is that you don't preserve the property value

Reply:
@mchaarawi it looks like it is setting a new ROOT_OID for the newly deserialized container when it calls dfs_cont_create, but instead it should use the same ones from the serialized container? Is that ok if the serialized container is not yet destroyed?

Reply:
correct for first question.
for second question, yes it's OK.. OIDs are per container, so it doesn't matter if the other container exists or no

Reply:
@mchaarawi ok, so it needs to be preserved. Maybe I am missing something but won't the ROOTS OID be overwritten if I pass it into dfs_cont_create? Since a new one is always generated?

Reply:
yes it will be overwritten in this case.. hmm
so we either:
1) update DFS to not overwrite in case that property is passed in (update dfs_cont_create)
2) set the prop again after the container is created (can be racy here)
3) make sure we set the oclass when calling dfs_cont_create, and thus we guarantee the OIDs will be the same.

thinking about this more, how do you preserver the chunk size and DFS mode in the serialization/deserialization? this could help in resolving this.

Reply:
again Im ok to get this in, and we can fix this issue in a different PR.
for now we don't preserve the OID settings through the serialize/deserialize

Reply:
@mchaarawi ok, I am comfortable with this getting merged now. I initially wanted to update it if it would cause an error for deserializing POSIX containers. It looks like there is a conflict in the changeling of the spec file again? Do I have to push and go through CI again?

Reply:
Please create a ticket to track the issue. this is a bug that needs tracking before this is usable.

you definitely need to rebase.
as to wait for CI or no, im fine to not wait for test results, but should build at least.
but you need a gatekeeper approval / forcelanding to do this.

Reply:
@mchaarawi created a ticket here https://jira.hpdd.intel.com/browse/DAOS-8000 and I've updated to latest master and fixed the merge conflict in changelog.